pkg/workload/schemachange/operation_generator.go

```diff
@@ -0,0 +1,1132 @@
+			stmt, err = og.dropColumn(tx)
+
+		case dropColumnDefault:
+			stmt, err = og.dropColumnDefault(tx)
+
+		case dropColumnNotNull:
+			stmt, err = og.dropColumnNotNull(tx)
+
+		case dropColumnStored:
+			stmt, err = og.dropColumnStored(tx)
+
+		case dropConstraint:
+			stmt, err = og.dropConstraint(tx)
+
+		case dropIndex:
+			stmt, err = og.dropIndex(tx)
+
+		case dropSequence:
+			stmt, err = og.dropSequence(tx)
+
+		case dropTable:
+			stmt, err = og.dropTable(tx)
+
+		case dropView:
+			stmt, err = og.dropView(tx)
+
+		case dropSchema:
+			stmt, err = og.dropSchema(tx)
+
+		case renameColumn:
+			stmt, err = og.renameColumn(tx)
+
+		case renameIndex:
+			stmt, err = og.renameIndex(tx)
+
+		case renameSequence:
+			stmt, err = og.renameSequence(tx)
+
+		case renameTable:
+			stmt, err = og.renameTable(tx)
+
+		case renameView:
+			stmt, err = og.renameView(tx)
+
+		case setColumnDefault:
+			stmt, err = og.setColumnDefault(tx)
+
+		case setColumnNotNull:
+			stmt, err = og.setColumnNotNull(tx)
+
+		case setColumnType:
+			stmt, err = og.setColumnType(tx)
+
+		case insertRow:
+			stmt, err = og.insertRow(tx)
+
+		case validate:
+			stmt, err = og.validate(tx)
+		}
+
+		// TODO(spaskob): use more fine-grained error reporting.
+		if stmt == "" || errors.Is(err, pgx.ErrNoRows) {
+			log.WriteString(fmt.Sprintf("NOOP: %s -> %v\n", op, err))
+			continue
+		}
+		return stmt, log.String(), expectedErrorCode, err
+	}
+}
+
+func (og operationGenerator) addColumn(tx *pgx.Tx) (string, *pgcode.Code, error) {
+
+	tableName, err := og.randTable(tx, og.pctExisting(true), "")
+	if err != nil {
+		return "", nil, err
+	}
+	tableExists, err := tableExists(tx, tableName)
+	if err != nil {
+		return "", nil, err
+	}
+	if !tableExists {
+		return fmt.Sprintf(`ALTER TABLE %s ADD COLUMN "%s" STRING`, tableName, "IrrelevantColumnName"), &pgcode.UndefinedTable, nil
+	}
+
+	columnName, err := og.randColumn(tx, *tableName, og.pctExisting(false))
+	if err != nil {
+		return "", nil, err
+	}
+
+	typ, err := og.randType(tx, og.pctExisting(true))
+	if err != nil {
+		return "", nil, err
+	}
+
+	def := &tree.ColumnTableDef{
+		Name: tree.Name(columnName),
+		Type: typ,
+	}
+	def.Nullable.Nullability = tree.Nullability(rand.Intn(1 + int(tree.SilentNull)))
+
+	columnExists, err := columnExistsOnTable(tx, tableName, columnName)
+	typeExists, typErr := typeExists(tx, typ)
+	if err != nil || typErr != nil {
+		return "", nil, err
+	}
+
+	var errCode *pgcode.Code
+	if columnExists {
+		if !typeExists {
+			errCode = &pgcode.UndefinedObject
+		} else if def.Nullable.Nullability == tree.NotNull {
+			colContainsNull, err := colContainsNull(tx, tableName, columnName)
+			if err != nil {
+				return "", nil, err
+			}
+			if colContainsNull {
+				errCode = &pgcode.NotNullViolation
+			} else {
+				errCode = &pgcode.DuplicateColumn
+			}
+		} else {
+			errCode = &pgcode.DuplicateColumn
+		}
+	} else {
+		if !typeExists {
+			errCode = &pgcode.UndefinedObject
+		}
+	}
```

Comment:
When the `NotNull` constraint is present, the behaviour with returning either a `NotNullViolation` or `DuplicateColumn` is pretty unpredictable. I thought I had it correct here, but eventually this happened:
```
root@localhost:26257/schemachange> create table foo(t TIMESTAMPTZ);
CREATE TABLE
Time: 80ms total (execution 80ms / network 0ms)

root@localhost:26257/schemachange> insert into foo values ('2016-06-22 19:10:25-07');
INSERT 1
Time: 43ms total (execution 43ms / network 0ms)

root@localhost:26257/schemachange> ALTER TABLE foo ADD COLUMN t TIMESTAMPTZ NOT NULL;
ERROR: null value in column "t" violates not-null constraint
SQLSTATE: 23502
```

In a way this is good because it adds test coverage. The downsides would be having to figure out exactly what causes each error and having to put this level of detail into each op. In this case, it may even be nondeterministic. It might be a good idea to return set of errors instead of one and checking that the resultant error exists in the set.